<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Calls Copy Number Variants from Targeted Sequence Data</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>
<h1> Calls Copy Number Variants from Targeted Sequence Data
<img class="toplogo" src="../../../doc/html/logo.jpg" alt="[R logo]">
</h1>
<hr>
<div align="center">
<a href="../../../doc/html/packages.html"><img src="../../../doc/html/left.jpg" alt="[Up]" width="30" height="30" border="0"></a>
<a href="../../../doc/html/index.html"><img src="../../../doc/html/up.jpg" alt="[Top]" width="30" height="30" border="0"></a>
</div><h2>Documentation for package &lsquo;ExomeDepth&rsquo; version 1.1.5</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../doc/index.html">User guides, package vignettes and other documentation.</a></li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td width="25%"><a href="ExomeDepth-package.html">ExomeDepth-package</a></td>
<td>Read depth based CNV calls for exome DNA sequence data</td></tr>
<tr><td width="25%"><a href="ExomeDepth-class.html">AddAnnotations-method</a></td>
<td>Class 'ExomeDepth'</td></tr>
<tr><td width="25%"><a href="AnnotateExtra.html">AnnotateExtra</a></td>
<td>Add annotations to a ExomeDepth object</td></tr>
<tr><td width="25%"><a href="AnnotateExtra-methods.html">AnnotateExtra-method</a></td>
<td>Additional annotations for ExomeDepth objects</td></tr>
<tr><td width="25%"><a href="AnnotateExtra-methods.html">AnnotateExtra-methods</a></td>
<td>Additional annotations for ExomeDepth objects</td></tr>
<tr><td width="25%"><a href="CallCNVs.html">CallCNVs</a></td>
<td>Call CNV data from an ExomeDepth object.</td></tr>
<tr><td width="25%"><a href="ExomeDepth-class.html">CallCNVs-method</a></td>
<td>Class 'ExomeDepth'</td></tr>
<tr><td width="25%"><a href="Conrad.hg19.common.CNVs.html">Conrad.hg19.common.CNVs</a></td>
<td>Conrad et al common CNVs</td></tr>
<tr><td width="25%"><a href="count.everted.reads.html">count.everted.reads</a></td>
<td>Count the number of everted reads for a set of BAM files.</td></tr>
<tr><td width="25%"><a href="countBam.everted.html">countBam.everted</a></td>
<td>Counts everted reads from a single BAM file</td></tr>
<tr><td width="25%"><a href="countBamInGRanges.exomeDepth.html">countBamInGRanges.exomeDepth</a></td>
<td>Compute read count data from BAM files.</td></tr>
<tr><td width="25%"><a href="ExomeCount.html">ExomeCount</a></td>
<td>Example dataset for ExomeDepth</td></tr>
<tr><td width="25%"><a href="ExomeDepth-package.html">ExomeDepth</a></td>
<td>Read depth based CNV calls for exome DNA sequence data</td></tr>
<tr><td width="25%"><a href="ExomeDepth-class.html">ExomeDepth-class</a></td>
<td>Class 'ExomeDepth'</td></tr>
<tr><td width="25%"><a href="exons.hg19.html">exons.hg19</a></td>
<td>Positions of exons on build hg19 of the human genome</td></tr>
<tr><td width="25%"><a href="exons.hg19.X.html">exons.hg19.X</a></td>
<td>Positions of exons on build hg19 of the human genome and on chromosome X</td></tr>
<tr><td width="25%"><a href="genes.hg19.html">genes.hg19</a></td>
<td>Positions of genes on build hg19 of the human genome</td></tr>
<tr><td width="25%"><a href="get.power.betabinom.html">get.power.betabinom</a></td>
<td>Estimate the power to compare two beta-binomial distributions.</td></tr>
<tr><td width="25%"><a href="getBamCounts.html">getBamCounts</a></td>
<td>Get count data for multiple exomes</td></tr>
<tr><td width="25%"><a href="initialize-methods.html">initialize-method</a></td>
<td>~~ Methods for Function 'initialize' ~~</td></tr>
<tr><td width="25%"><a href="initialize-methods.html">initialize-methods</a></td>
<td>~~ Methods for Function 'initialize' ~~</td></tr>
<tr><td width="25%"><a href="plot.ExomeDepth.html">plot-method</a></td>
<td>Plotting function for ExomeDepth objects</td></tr>
<tr><td width="25%"><a href="plot.ExomeDepth.html">plot-methods</a></td>
<td>Plotting function for ExomeDepth objects</td></tr>
<tr><td width="25%"><a href="plot.ExomeDepth.html">plot.ExomeDepth</a></td>
<td>Plotting function for ExomeDepth objects</td></tr>
<tr><td width="25%"><a href="qbetabinom.html">qbetabinom</a></td>
<td>Quantile for betabin function</td></tr>
<tr><td width="25%"><a href="qbetabinom.ab.html">qbetabinom.ab</a></td>
<td>Quantile function for the beta-binomial distribution</td></tr>
<tr><td width="25%"><a href="select.reference.set.html">select.reference.set</a></td>
<td>Combine multiple samples to optimize the reference set in order to maximise the power to detect CNV.</td></tr>
<tr><td width="25%"><a href="show-methods.html">show-method</a></td>
<td>~~ Methods for Function 'show' ~~</td></tr>
<tr><td width="25%"><a href="show-methods.html">show-methods</a></td>
<td>~~ Methods for Function 'show' ~~</td></tr>
<tr><td width="25%"><a href="somatic.CNV.call.html">somatic.CNV.call</a></td>
<td>Call somatic variants between healthy and disease tissues.</td></tr>
<tr><td width="25%"><a href="TestCNV.html">TestCNV</a></td>
<td>Computes the Bayes Factor in favour of a CNV defined by position and type.</td></tr>
<tr><td width="25%"><a href="ExomeDepth-class.html">TestCNV-method</a></td>
<td>Class 'ExomeDepth'</td></tr>
<tr><td width="25%"><a href="viterbi.hmm.html">viterbi.hmm</a></td>
<td>Computes the Viterbi path for a hidden markov model</td></tr>
</table>
</body></html>
