library(biomaRt)
library(tidyr)
library(dplyr)
library(readxl)
library(R.utils)

args=commandArgs(asValues = TRUE)

# Define dir
outputDir <- args$outputDir
bedFile <- args$bedFile
panellsFile <- args$panellsFile

# make temp dir, results dir en graphs dir
tempDir <- file.path(outputDir, "temp")

if(!dir.exists(tempDir)) dir.create(tempDir)

# load bedfile
bedData <- read.delim(bedFile, header=FALSE)

# load ensamble
ensembl = useMart("ensembl",dataset="hsapiens_gene_ensembl")

# load panell names
panells <- c("mama", "colon", "melanoma", "endocri")

# run over panells to annotate genes included
for (panell in panells){
  
  # Read genes included in panell
  genes_panell <- read_excel(panellsFile, sheet = panell, col_names = FALSE)
  colnames(genes_panell) <- c("gene", "refseq_id")
  
  # annotate for given genes
  my_attribute <- c('chromosome_name',
                  'refseq_mrna',
                  'external_gene_name',
                  'strand',
                  'ensembl_transcript_id')
  
  #fetch gene info
  a <- getBM(attributes=my_attribute,
           filters = c('refseq_mrna'),
           values = list(refseq_mrna=genes_panell$refseq_id),
           mart = ensembl)

  #build attribute vector 2
  my_attribute2_exon <- c('exon_chrom_start',
                        'exon_chrom_end',
                        'rank',
                        'ensembl_transcript_id',
                        'cds_start', 'cds_end')

  #fetch exon info
  b <- getBM(attributes=my_attribute2_exon,
           filters = c('refseq_mrna'),
           values = list(refseq_mrna=genes_panell$refseq_id),
           mart = ensembl)
  
  # merge both annotation tables
  genes_info <- merge(a, b)
  genes_info <- genes_info[, c(2, 6, 7, 4, 3, 5, 8:10)]
  colnames(genes_info) <- c("chr", "start", "end", "gene", "nm","strand", "rank", "cds_start", "cds_end")
  write.table(genes_info, file.path(tempDir, "genes_info.bed"), sep="\t", row.names=FALSE, quote = FALSE, col.names = FALSE)  
  
  # Load bedFile and subset on genes in panell
  bedFilePanel <- subset(bedData, bedData$V4 %in% genes_panell$gene)
  colnames(bedFilePanel) <- c("chr", "start", "end", "gene")

  # Give ID to region in bed
  bedFilePanel$roiID <- 1:nrow(bedFilePanel)
  write.table(bedFilePanel, file.path(tempDir, "bedFilePanel.bed"), sep="\t", row.names=FALSE, quote = FALSE, col.names = FALSE)  
  
  # Intersect bedfile and annotated information
  setwd(tempDir)
  system(paste("bedtools intersect -wao -a bedFilePanel.bed -b genes_info.bed >", paste0(panell, "_annotated.bed")))
  system(paste("bedtools intersect -wao -a bedFilePanel.bed -b genes_info.bed -v >", paste0(panell, "_notpresent.bed")))
}

# Read annotated bed files for each panel
colonFile <- read.delim(file.path(tempDir, "colon_annotated.bed"), header = FALSE)[c(1:5,10:14)]
mamaFile <- read.delim(file.path(tempDir, "mama_annotated.bed"), header = FALSE)[c(1:5,10:14)]
melanomaFile <- read.delim(file.path(tempDir, "melanoma_annotated.bed"), header = FALSE)[c(1:5,10:14)]
endocriFile <- read.delim(file.path(tempDir, "endocri_annotated.bed"), header = FALSE)[c(1:5,10:14)]

newnames <- c("chr", "start", "end", "gene","roi_id", "nm","strand", "rank", "cds_start", "cds_end")
names(colonFile) <- names(mamaFile) <- names(melanomaFile) <- names(endocriFile) <- newnames

# Delete rois outside NM transcript
delRoi <- function(dataFile){
  if(length(which(dataFile$nm == ".")) > 0){
    dataFile <- dataFile[- which(dataFile$nm == "."), ]
  }
  return(dataFile)
}

colonFile <- delRoi(colonFile)
mamaFile <- delRoi(mamaFile)
melanomaFile <- delRoi(melanomaFile)
endocriFile <- delRoi(endocriFile)

# Add new start and end for figure of report
NewExonStart <- function(file){
  file <- file[order(file$gene, as.numeric(file$rank)), ]
  file$exonpb <- file$end - file$start
  file$intropb <- NA
  file$newstart <- NA
  file$newend <- NA
  genes <- unique(file$gene)
  for(gene in genes){
    a <- which(file$gene == gene & file$strand == 1)
    if(length(a) > 0){
      file$intropb[a] <- abs(round((file$end[a] - lead(file$start)[a]) / 1000, 0))
    }
    b <- which(file$gene == gene & file$strand == -1)
    if(length(b) > 0){
      file$intropb[b] <- round((file$start[b] - lead(file$end)[b]) / 1000, 0)
    }
    n <- which(file$gene == gene)
    newcoor <- c(0, head(cumsum(c(rbind(file$exonpb[n], file$intropb[n]))), -1))
    file$newstart[n] <- newcoor[seq(from = 1, to = length(newcoor), by = 2)]
    file$newend[n] <- newcoor[seq(from = 2, to = length(newcoor), by = 2)]
  }
  return(file)
}

colonFile <- NewExonStart(colonFile)
mamaFile <- NewExonStart(mamaFile)
endocriFile <- NewExonStart(endocriFile)
melanomaFile <- NewExonStart(melanomaFile) 

# rbind all annotated files and write out table
annotatedFile <- data.frame(rbind(colonFile, mamaFile, melanomaFile, endocriFile), panell = c(rep("colon", nrow(colonFile)), rep("mama", nrow(mamaFile)), rep("melanoma", nrow(melanomaFile)), rep("endocri", nrow(endocriFile)))) 
write.table(annotatedFile, file.path(outputDir, "annotatedFile.bed"), sep="\t", row.names=FALSE, quote = FALSE, col.names = TRUE)  

