
library(R.utils)

args=commandArgs(asValues = TRUE)

library(panelcn.mops)
library(GenomicRanges)

#bam_file=args$bams
#bedfile=args$bed
#fasta=args$fasta
#exonfile=args$exon
#output=args$out

bam_file="/home/robin/Documents/Project/Samples/example/bam_ex"                                                                   #location of bam files; can be a directory containing only bam files to be processed or the name of a file containing a list of bam files to be processed.
bedfile="/home/robin/Documents/Project/Samples/bedfiles/ICR96_hg38.bed"                                                                       #name of bed file
fasta="/home/robin/Documents/Project/Samples/hg38/hg38.fa"                                                                     #name of fasta file
exonfile="/home/robin/Documents/Project/Samples/bedfiles/exons_hg38.bed"
output="/home/robin/Documents/Project/Results/panelcnmops"

exon.bed<-read.table(paste(exonfile))                                             
exons <- exon.bed$V5

out <- strsplit(exons, "_")
do.call(rbind, out)

# Input
countWindow <- getWindows(bedfile)

