<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Copy number variant detection from exome sequencing read depth</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Copy number variant detection from exome sequencing read depth
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;exomeCopy&rsquo; version 1.36.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../doc/index.html">User guides, package vignettes and other documentation.</a></li>
<li><a href="../NEWS">Package NEWS</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="exomeCopy-package.html">exomeCopy-package</a></td>
<td>Copy number variant detection from exome sequencing read depth</td></tr>
<tr><td style="width: 25%;"><a href="compileCopyCountSegments.html">compileCopyCountSegments</a></td>
<td>Compile segments across samples</td></tr>
<tr><td style="width: 25%;"><a href="copyCountSegments.html">copyCountSegments</a></td>
<td>Segments of identical copy count from exomeCopy</td></tr>
<tr><td style="width: 25%;"><a href="countBamInGRanges.html">countBamInGRanges</a></td>
<td>Count reads from BAM file in genomic ranges</td></tr>
<tr><td style="width: 25%;"><a href="exomeCopy.html">exomeCopy</a></td>
<td>Fit the exomeCopy or exomeCopyVar model to the observed counts.</td></tr>
<tr><td style="width: 25%;"><a href="ExomeCopy-class.html">ExomeCopy-class</a></td>
<td>Class "ExomeCopy"</td></tr>
<tr><td style="width: 25%;"><a href="exomecounts.html">exomecounts</a></td>
<td>Sample counts from 16 exome sequencing samples from 1000 Genomes Project</td></tr>
<tr><td style="width: 25%;"><a href="generateBackground.html">generateBackground</a></td>
<td>Generate median background read depth</td></tr>
<tr><td style="width: 25%;"><a href="getGCcontent.html">getGCcontent</a></td>
<td>Get the GC content of target ranges from a reference FASTA file</td></tr>
<tr><td style="width: 25%;"><a href="negLogLike.html">negLogLike</a></td>
<td>Generalized negative log likelihood and Viterbi algorithms</td></tr>
<tr><td style="width: 25%;"><a href="ExomeCopy-class.html">plot-method</a></td>
<td>Class "ExomeCopy"</td></tr>
<tr><td style="width: 25%;"><a href="plot.exomeCopyObject.html">plot.ExomeCopy</a></td>
<td>Plot function for exomeCopy</td></tr>
<tr><td style="width: 25%;"><a href="plotCompiledCNV.html">plotCompiledCNV</a></td>
<td>Plot compiled CNV segments for one sequence/chromosome</td></tr>
<tr><td style="width: 25%;"><a href="ExomeCopy-class.html">show-method</a></td>
<td>Class "ExomeCopy"</td></tr>
<tr><td style="width: 25%;"><a href="subdivideGRanges.html">subdivideGRanges</a></td>
<td>Subdivide ranges of a GRanges object into nearly equal width ranges</td></tr>
<tr><td style="width: 25%;"><a href="negLogLike.html">viterbiPath</a></td>
<td>Generalized negative log likelihood and Viterbi algorithms</td></tr>
</table>
</body></html>
